File: src/haskell/src/Cat/NaturalTransformation.hs
-- | Natural transformation -- morphism between functors.
--
-- Mathematical definition (Stacks Project 0015):
-- Given functors \(F, G : \mathcal{C} \to \mathcal{D}\), a natural transformation
-- \(\eta : F \Rightarrow G\) is a family of morphisms
-- \(\eta_a : F(a) \to G(a)\) for each object \(a\) in \(\mathcal{C}\),
-- such that for every morphism \(f : a \to b\) in \(\mathcal{C}\),
-- the naturality square commutes:
-- \(G(f) \circ \eta_a = \eta_b \circ F(f)\).
module Cat.NaturalTransformation
  ( -- * Natural transformations
    NatTrans(..)
    -- * Vertical composition
  , idNat
  , vertComp
  ) where

import Prelude hiding (id, (.))
import Data.Kind (Type)

import Cat.Category (Category(..))

--------------------------------------------------------------------------------
-- Natural transformations
--------------------------------------------------------------------------------

-- | A natural transformation \(\eta : F \Rightarrow G\) between functors
-- @f@ and @g@ landing in a category with morphisms @cat2@.
--
-- The naturality condition is not enforced by the type system but is a
-- law that instances must satisfy:
--
-- @G(h) \`compose\` component eta = component eta \`compose\` F(h)@
--
-- for all morphisms @h@ in the source category.
--
-- Stacks Project 0015, Definition 3.4.
newtype NatTrans (cat2 :: k2 -> k2 -> Type) (f :: k1 -> k2) (g :: k1 -> k2) = NatTrans
  { component :: forall (a :: k1). cat2 (f a) (g a)
    -- ^ The component at object @a@: a morphism \(\eta_a : F(a) \to G(a)\).
  }

--------------------------------------------------------------------------------
-- Vertical composition
--------------------------------------------------------------------------------

-- | The identity natural transformation \(\mathrm{id}_F : F \Rightarrow F\).
-- Each component is the identity morphism: \((\mathrm{id}_F)_a = \mathrm{id}_{F(a)}\).
--
-- nLab: identity natural transformation.
idNat :: Category cat2 => NatTrans cat2 f f
idNat = NatTrans id

-- | Vertical composition of natural transformations.
-- Given \(\alpha : F \Rightarrow G\) and \(\beta : G \Rightarrow H\),
-- \((\beta \bullet \alpha)_a = \beta_a \circ \alpha_a\).
--
-- Stacks Project 0016.
vertComp :: Category cat2 => NatTrans cat2 g h -> NatTrans cat2 f g -> NatTrans cat2 f h
vertComp (NatTrans beta) (NatTrans alpha) = NatTrans (compose beta alpha)

File: src/julia/src/NaturalTransformation.jl
# Natural transformation -- morphism between functors.
#
# Mathematical definition (Stacks Project 001I):
# Given functors F, G : C -> D, a natural transformation eta : F => G
# is a family of morphisms eta_a : F(a) -> G(a) for each object a in C,
# such that for every morphism f : a -> b in C, the naturality square commutes:
# G(f) . eta_a = eta_b . F(f).
#
# GATlab encodes this as a multi-sorted theory extending the "two parallel
# functors" setup. We have a source category C, a target category D, two
# functors F and G (object and morphism maps for each), and a component
# family eta indexed by objects of C, valued in morphisms of D.
#
# This follows the pattern of Catlab's ThCategory2, where 2-cells are typed
# as Hom2(dom::Hom(A,B), codom::Hom(A,B))::TYPE -- a sort dependent on
# 1-morphisms. Here, components are morphisms in D dependent on objects of C.
#
# Stacks Project 001I; nLab: natural+transformation.

using GATlab

"""
    ThNaturalTransformation

The theory of a natural transformation eta : F => G between functors F, G : C -> D.

Extends the two-functor setup with:
- `component(a::ObC)` -- the component eta_a : HomD(fob(a), gob(a))

Axioms:
- Naturality: for all f : a -> b in C,
  composeD(fhom(f), component(b)) == composeD(component(a), ghom(f))

This theory inlines the source/target category axioms and both functors'
laws. The naturality condition is the key additional axiom.

Stacks Project 001I; nLab: natural+transformation.
"""
@theory ThNaturalTransformation begin
    # Source category C
    ObC::TYPE
    HomC(dom::ObC, codom::ObC)::TYPE
    idC(a::ObC)::HomC(a, a)
    composeC(f::HomC(a, b), g::HomC(b, c))::HomC(a, c) ⊣ [a::ObC, b::ObC, c::ObC]

    # Source category laws
    composeC(f, idC(b)) == f ⊣ [a::ObC, b::ObC, f::HomC(a, b)]
    composeC(idC(a), f) == f ⊣ [a::ObC, b::ObC, f::HomC(a, b)]
    composeC(composeC(f, g), h) == composeC(f, composeC(g, h)) ⊣
        [a::ObC, b::ObC, c::ObC, d::ObC, f::HomC(a, b), g::HomC(b, c), h::HomC(c, d)]

    # Target category D
    ObD::TYPE
    HomD(dom::ObD, codom::ObD)::TYPE
    idD(a::ObD)::HomD(a, a)
    composeD(f::HomD(a, b), g::HomD(b, c))::HomD(a, c) ⊣ [a::ObD, b::ObD, c::ObD]

    # Target category laws
    composeD(f, idD(b)) == f ⊣ [a::ObD, b::ObD, f::HomD(a, b)]
    composeD(idD(a), f) == f ⊣ [a::ObD, b::ObD, f::HomD(a, b)]
    composeD(composeD(f, g), h) == composeD(f, composeD(g, h)) ⊣
        [a::ObD, b::ObD, c::ObD, d::ObD, f::HomD(a, b), g::HomD(b, c), h::HomD(c, d)]

    # Functor F : C -> D
    fob(a::ObC)::ObD
    fhom(f::HomC(a, b))::HomD(fob(a), fob(b)) ⊣ [a::ObC, b::ObC]

    # F preserves identity and composition
    fhom(idC(a)) == idD(fob(a)) ⊣ [a::ObC]
    fhom(composeC(f, g)) == composeD(fhom(f), fhom(g)) ⊣
        [a::ObC, b::ObC, c::ObC, f::HomC(a, b), g::HomC(b, c)]

    # Functor G : C -> D
    gob(a::ObC)::ObD
    ghom(f::HomC(a, b))::HomD(gob(a), gob(b)) ⊣ [a::ObC, b::ObC]

    # G preserves identity and composition
    ghom(idC(a)) == idD(gob(a)) ⊣ [a::ObC]
    ghom(composeC(f, g)) == composeD(ghom(f), ghom(g)) ⊣
        [a::ObC, b::ObC, c::ObC, f::HomC(a, b), g::HomC(b, c)]

    # Natural transformation component
    component(a::ObC)::HomD(fob(a), gob(a))

    # Naturality condition (diagrammatic order):
    # composeD(fhom(f), component(b)) == composeD(component(a), ghom(f))
    # i.e., (fhom(f) then eta_b) == (eta_a then ghom(f))
    composeD(fhom(f), component(b)) == composeD(component(a), ghom(f)) ⊣
        [a::ObC, b::ObC, f::HomC(a, b)]
end
